exe/Daemon.hs

```diff
@@ -0,0 +1,485 @@
+-- materialises the state of remote IPFS machines on the users PC, and
+-- writes to those IPFS machines the user is an owner of.
+--
+-- See
+-- <https://github.com/oscoin/radicle/blob/master/rfcs/0003-radicle-daemon.rst
+-- the RFC>.
+module Daemon where
+
+import           Protolude hiding (fromStrict, option, poll)
+
+import qualified Data.Aeson as A
+import           Data.ByteString.Lazy (fromStrict)
+import qualified Data.Map.Strict as Map
+import qualified Data.Sequence as Seq
+import qualified Data.Time.Clock.System as Time
+import           Network.Wai.Handler.Warp (run)
+import           Options.Applicative
+import           Servant
+import           System.Directory (doesFileExist)
+
+import           Radicle.Daemon.Common
+import qualified Radicle.Internal.UUID as UUID
+
+import           Radicle hiding (Env)
+import           Radicle.Daemon.HttpApi
+import           Radicle.Daemon.Ipfs
+import qualified Radicle.Internal.CLI as Local
+import qualified Radicle.Internal.ConcurrentMap as CMap
+
+
+-- TODO(james): Check that the IPFS functions are doing all the
+-- necessary pinning.
+
+-- * Types
+
+data MachineError
+  = InvalidInput (LangError Value)
+  | IpfsError IpfsError
+  | AckTimeout
+  | DaemonError Text
+  | MachineAlreadyCached
+  | MachineNotCached
+
+data Error
+  = MachineError MachineId MachineError
+  | CouldNotCreateMachine Text
+
+displayError :: Error -> (Text, [(Text,Text)])
+displayError = \case
+  CouldNotCreateMachine e -> ("Could not create IPFS machine", [("error", e)])
+  MachineError id e -> let mid = ("id", getMachineId id) in
+    case e of
+      InvalidInput err -> ("Invalid radicle input", [mid, ("error", renderCompactPretty err)])
+      DaemonError err -> ("Internal error", [mid, ("error", err)])
+      IpfsError e' -> case e' of
+        IpfsDaemonError err -> ("There was an error using IPFS", [mid, ("error", toS (displayException err))])
+        InternalError err -> ("Internal error", [mid, ("error", err)])
+        NetworkError err -> ("There was an error communicating with the IPFS daemon", [mid, ("error", err)])
+      AckTimeout -> ("The writer appears to be offline", [mid])
+      MachineAlreadyCached -> ("Tried to add already cached machine", [mid])
+      MachineNotCached -> ("Machine was not found in cache", [mid])
+
+logDaemonError :: MonadIO m => Error -> m ()
+logDaemonError (displayError -> (m, xs)) = logErr m xs
+
+type Follows = Map MachineId ReaderOrWriter
+
+-- * Main
+
+data Opts = Opts
+  { port       :: Int
+  -- TODO(james): temporary, just for testing.
+  , filePrefix :: Text
+  }
+
+opts :: Parser Opts
+opts = Opts
+    <$> option auto
+        ( long "port"
+       <> help "daemon port"
+       <> metavar "PORT"
+       <> showDefault
+       <> value 8909
+        )
+    <*> strOption
+        ( long "filePrefix"
+       <> help "file prefix"
+       <> metavar "PREFIX"
+       <> showDefault
+       <> value ""
+        )
+
+type FollowFileLock = MVar ()
```

Comment:
How bad is it if two different daemons are started? Do they notice one another and quit? If not, and clobbering the file is serious, it may be worth using `flock` (via e.g. [`filelock`](http://hackage.haskell.org/package/filelock-0.1.1.2/docs/System-FileLock.html)) since MVars only work within one process.

Reply:
This was just to avoid separate threads in one daemon from clobbering each other. I will use `filelock` to prevent two daemons from starting up (it didn't yet because it was nice to run two for testing).

Reply:
Added to tech-debt.